### ISSUE TITLE
Euclidean Distance Transform 

### ISSUE THREAD

**Issue by ahatamiz** [tensorflow/tensorflow#24410]
Given the importance and wide use of the distance transform of binary images in different computer vision applications, I wonder if it would be possible to add a Euclidean distance transform with gradient support. 

There exists very nice O(n) solutions already implemented in C/C++. For example. Opencv has implemented this [paper](https://www.realestatetrading.com/storage/documents/Borgefors86__Work.pdf) [here](https://docs.opencv.org/4.0.0/d7/d1b/group__imgproc__misc.html#ga8a0b7fdfcb7a13dde018988ba3a43042). 

Here's another fast [implementation](https://diplib.github.io/diplib-docs/group__distance.html#ga5feb444b79b00710cdd5b371d55fd221) in PyDIP. 

This would be a tremendously useful feature. Thanks!

**Comment by skye** [comment]
@martinwicke @ebrevdo @rmlarsen do you know if there's a TensorFlower who'd be interested in this?

**Comment by FrankSD** [comment]
I am  a beginner contributer?Is this suitable  for  first timer?Which  lebel  for  first time or beginner in  tensorflow?

**Comment by skye** [comment]
Unfortunately I do not think this is a good project for a beginner. This is probably best for someone with experience writing TensorFlow ops.

**Comment by skye** [comment]
Hi @Squadrick, I'm trying to find someone on the TF team who would be a good reviewer for this (and also give a better estimate of how much work this is, etc.). In the meantime though, you could start development in https://github.com/tensorflow/addons if you like.

**Comment by martinwicke** [comment]
@sofienbouaziz would be a good reviewer for an op. I would suggest starting in tensorflow/addons.

@seanpmorgan FYI.

**Comment by Squadrick** [comment]
I'll get started right away.

**Comment by sendeb** [comment]
Are there any updates on this? This will be an amazing feature when available!

**Comment by ahatamiz** [comment]
Following :)

**Comment by Squadrick** [comment]
Hey, guys. Should be done within this week. I'll update this thread once I open a PR.

**Comment by ahatamiz** [comment]
Thank you so much for your great work.

**Comment by sendeb** [comment]
Is there an update on this feature by any chance? Thanks!

**Comment by Squadrick** [comment]
@skye Sorry about the delay, I've created a PR with a barebones implementation. I'll add faster algorithms and test cases over the next few hours.

**Comment by ahatamiz** [comment]
Great job !  How would we be able to access this custom_op ?

**Comment by Squadrick** [comment]
@ahatamiz 

The changes haven't been merged into the main repository yet, so clone my [repository](https://github.com/Squadrick/addons), and switch to this branch: `euclidean-distance-transform`.

[Here's](https://github.com/Squadrick/addons/blob/euclidean-distance-transform/README.md#installation) the build guide for tensorflow_addons.

Then you can use the transform as such:
```
from tensorflow_addons.custom_ops.image import euclidean_dist_transform

x = ... # binary image
y = euclidean_dist_transform(x)
```

Note:

1. `x` must be a binary image (0 and 1 only)
2. `x` must be of type `tf.uint8`. Use `tf.cast` if it isn't.
3. `x` must be of shape `[N, H, W, 1]` or `[H, W, 1]`
4. The dtype of `y` can be set like this: `y = euclidean_dist_transform(x, dtype=tf.float32)`. `tf.float32` is the default.
5. The current algorithm is of quadratic complexity `O(N^2)` where `N` is the number of pixels. I'm working on the faster `O(N)` algorithm right now, and I'll push it soon.
6. The Op is currently CPU only. 
7. The Op is not differentiable, and running `tf.gradients(y)` will return `[None]`.

Let me know if you find any problems or issues with the Op.

**Comment by ahatamiz** [comment]
@Squadrick  

That's amazing ! this op will be tremendously helpful for many researchers and I am sure people will come to appreciate its importance soon ! 

The only concern though is the fact that Op is non-differentiable. It would be great if we could have gradients defined for this op.

In many applications, this op will be needed for connecting two layers in a sense that it initializes the second layer with the output of the first layer. In essense, both of the layers have learn-able parameters thus not being able to back-propagate the of error through this op would limit its use.

**Comment by skye** [comment]
@Squadrick awesome! Once it's ready with full test cases etc., maybe ping @sofienbouaziz to take a look.

**Comment by Squadrick** [comment]
@skye Sure thing.

@ahatamiz Couldn’t really figure out what the gradient of the operation would be, considering the fact that the function is defined on binary images. If you could help out with that part, I can implement fairly easily.

**Comment by sofienbouaziz** [comment]
@Squadrick Thanks for the contribution! Just quickly parsing through the code I have the same comment than @ahatamiz that this op is non-differentiable and it would be great if we could back-propagate through it. The algorithm currently used is O(n^2)  is the goal to then implement a O(n) algorithm?

**Comment by Squadrick** [comment]
@sofienbouaziz I'm working on the `O(n)` algorithm right now, I'll push it within a day. 

Like I mentioned before, I'm not sure how this op can be differentiable considering the fact that the op is defined on binary images.

**Comment by sofienbouaziz** [comment]
@Squadrick Awesome for the O(n) version! Let me think about how to differentiate this op during the week-end. Just out of curiosity do you think it would be possible to develop this op purely in python?

**Comment by drebain** [comment]
@sofienbouaziz I think @Squadrick is right that gradients make no sense for a binary input. However, there is an [generalization for arbitrary functions](http://www.theoryofcomputing.org/articles/v008a019/v008a019.pdf) (that you and @ataiya [implemented?](https://github.com/ataiya/dtform)) that if I understand correctly should be differentiable. If there was an op for this you would also get the binary version for free by composing it with an indicator function.

I'm not sure if the generalized version would do exactly what @ahatamiz wants, but it would be very nice to have an op for.

**Comment by drebain** [comment]
As an experiment I tried [implementing a simple version of the generalized transform in python](https://colab.research.google.com/drive/1-Fykc6FVi3pUtFx7spP3OAdn267TKwbD), and it is indeed differentiable. With a careful choice of indicator function it seems you can even "learn" a binary image by supervising its distance transform, so it might be suitable for use as a layer in a network as @ahatamiz suggested.

Below is an example showing the gradients. Here, the mask is learned from an L2 loss between the DT of the mask and the DT of the ground truth mask.

![smiley](https://user-images.githubusercontent.com/5511462/54079136-8aaa5800-42a4-11e9-9903-30bca37fce31.gif)

**Comment by ahatamiz** [comment]
Thanks @Squadrick for you great work on this and thanks @sofienbouaziz for taking a look into this. Also thanks for your thoughtful contribution to the discussion @drebain  

@Squadrick, you are right in the sense that the op may not be differentiable if we were to utilize on binary images. However,  what I had in mind is very close the implementation by @drebain. One can pass the output of sigmoid, before thresholding, to a (generalized) distance transform layer and still allow for gradient flow. This is useful since in most vision applications, the output of sigmoid (or softmax) can be fed into this layer for more interesting operation down-stream. 

The implementation by @drebain is definitely correct, but unfortunately it runs out of memory on bigger images ( like 256 by 256 ) on a GPU like Titan XP and also is not as fast, due to its O(n^2) complexity.

**Comment by drebain** [comment]
According to its documentation, the default O(n) L2 implementation in OpenCV uses the multi-pass algorithm based on the generalized transform, so if @Squadrick has implemented that, then I think all that would need to be done is to expose a version of the op that doesn't apply the thresholding indicator function.

**Comment by ahatamiz** [comment]
> @ahatamiz
> 
> The changes haven't been merged into the main repository yet, so clone my [repository](https://github.com/Squadrick/addons), and switch to this branch: `euclidean-distance-transform`.
> 
> [Here's](https://github.com/Squadrick/addons/blob/euclidean-distance-transform/README.md#installation) the build guide for tensorflow_addons.
> 
> Then you can use the transform as such:
> 
> ```
> from tensorflow_addons.custom_ops.image import euclidean_dist_transform
> 
> x = ... # binary image
> y = euclidean_dist_transform(x)
> ```
> Note:
> 
> 1. `x` must be a binary image (0 and 1 only)
> 2. `x` must be of type `tf.uint8`. Use `tf.cast` if it isn't.
> 3. `x` must be of shape `[N, H, W, 1]` or `[H, W, 1]`
> 4. The dtype of `y` can be set like this: `y = euclidean_dist_transform(x, dtype=tf.float32)`. `tf.float32` is the default.
> 5. The current algorithm is of quadratic complexity `O(N^2)` where `N` is the number of pixels. I'm working on the faster `O(N)` algorithm right now, and I'll push it soon.
> 6. The Op is currently CPU only.
> 7. The Op is not differentiable, and running `tf.gradients(y)` will return `[None]`.
> 
> Let me know if you find any problems or issues with the Op.

@Squadrick . I cloned the branch you mentioned and built from the source, but it seems like we are missing a file: 

`/addons-euclidean-distance-transform/tensorflow_addons/custom_ops/image/python/_image_ops.so  cannot open shared object file: No such file or directory`

Also, as @drebain  mentioned, I think it would be helpful to also allow for feeding logits before sigmoid or softmax and not just binary images ( for a back-propagating error ) . Currently, this layer seems to only accept binary images.

**Comment by seanpmorgan** [comment]
@ahatamiz I haven't gotten to fully review the PR, but I recognize that error and you likely just need to move out of the src directory after pip install. The .so files aren't built until bazel compile and your current working directory does not contain them

**Comment by ahatamiz** [comment]
@seanpmorgan Yes. You were right and it worked. 

Also, it seems like this layer requires TensorFlow version >= 2. Would it be possible to allow older versions like 1.9 for now ?  @Squadrick

**Comment by ahatamiz** [comment]
> @ahatamiz I haven't gotten to fully review the PR, but I recognize that error and you likely just need to move out of the src directory after pip install. The .so files aren't built until bazel compile and your current working directory does not contain them

Thanks @seanpmorgan . However I also have an issue with version incompatblity as only Tensorflow 2.0.0 is allowed. I have not yet transferred everything to Tensorflow 2.0.0 and was hoping to get it working on older versions.

**Comment by seanpmorgan** [comment]
@ahatamiz Ah, thats a bit of a problem. Addons is only scoped to be TF 2.x+ and it'll take a lot of convincing to get us to reconsider that. You can however make it work on your own fork of the repo by altering this line:
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/__init__.py#L54

**Comment by sofienbouaziz** [comment]
@drebain Very cool colab! Indeed the [code](https://github.com/ataiya/dtform) you are referring to is an old implementation I did of this particular [paper](http://www.theoryofcomputing.org/articles/v008a019/v008a019.pdf) and as you pointed out it is differentiable. Yes, you are right that the gradient doesn't make sense in the case of binary image, however I was thinking we could do something similar than what has been done for quantization where the forward pass would be computed on the binary image and the backward pass on some sort of relaxation.

**Comment by sofienbouaziz** [comment]
I am closing this issue as the Euclidean Distance Transform op was merged. If updates are needed let's reopen one with the necessary details. 
@Squadrick Thanks for this very useful op.

**Comment by ahatamiz** [comment]
Thanks you @Squadrick  and @sofienbouaziz . 

Is it also possible to provide GPU support for this op ?

**Comment by Squadrick** [comment]
@ahatamiz GPU support for TensorFlow Addons is currently WIP. Ref [#294](https://github.com/tensorflow/addons/pull/294). Once this is merged, I'll add GPU support.

**Comment by emckenzi123** [comment]
Hi,
I just want to confirm.  Can the current implementation of the TF addons Euclidean Distance Transform be used in a loss function?  I noticed the discussion above on the issue of gradients when the distance transform is applied on binary images.  Was this resolved?

**Comment by Squadrick** [comment]
@emckenzi123 No. The current implementation of euclidean distance transform does not have a gradient defined.

**Comment by drebain** [comment]
@emckenzi123 If you do not need to run on large images, the code in my notebook from above ( https://colab.research.google.com/drive/1-Fykc6FVi3pUtFx7spP3OAdn267TKwbD ) might work as a loss function. You do need to allow arbitrary real values for pixels (not just binary) in order for the gradients of a distance transform to be meaningful though.

**Comment by emckenzi123** [comment]
@drebain I tried your posted method (thank you for making that available) however I ran into an OOM error since my images are large.  I looked into the [paper](http://www.theoryofcomputing.org/articles/v008a019/v008a019.pdf) you linked to, and found a python version [here](https://github.com/hbristow/distance-transform).  It uses Cython, and I naively tried to just use this code in a training loop's loss, without success (no training of weights).  I went through and tried to replace ops with tensorflow ops to see if I could get something, but that just ended up being really slow.  I should note that if I use the Cython code outside of tensorflow, it's super fast.  I could try to keep working on this, or do you guys think it's actually non-trivial to turn the linked python code into something TF can use in a loss function?

**Comment by drebain** [comment]
@emckenzi123 I haven't used Cython and TF together before, but I think you could write a wrapper around the code you linked that manually defines the gradient using [tf.py_function](https://www.tensorflow.org/api_docs/python/tf/py_function) and [tf.custom_gradient](https://www.tensorflow.org/api_docs/python/tf/custom_gradient). If I'm reading correctly that the code gives you the index of the pixel corresponding to the minimum of d(p,q)+f(q) you should be able to compute the gradient. That would let you keep the Cython implementation I think, while still being differentiable.

**Comment by emckenzi123** [comment]
Thank you so much for your help.  I defined the gradient to be 2 * sqrt(distance transform), since the derivative of (p-q)**2 is 2 * (p-q).  
In case someone in the future runs into something similar, I have used @drebain 's previous colab notebook as a template to demonstrate the code for this.  The link is [here](https://colab.research.google.com/drive/13IxLV__fl20fWa3nCBruuYTcDH0-kwM5?usp=sharing)

**Comment by jacksonloper** [comment]
Needed a tensorflow op that did the generalized edt (i.e. `out[p] = min_q (||q-p||^2 + in[q])`) .  Implemented here, CPU/GPU, float32/float64.  

https://github.com/jacksonloper/tensorflow-felzenszwalb-edt

Wheel here

https://github.com/jacksonloper/tensorflow-felzenszwalb-edt/releases/download/v0.0.1/tensorflow_felzenszwalb_edt-0.0.1-py37-none-linux_x86_64.whl

Caveat emptor, haven't tested it on any platforms besides my own.  No idea how it compares in speed to the (non-differentiable) morphological edt.  

I guess if it seems like there is sufficient interest, I could package it up and submit PR.  Dunno what common practice is here (e.g. if 2000 people have downloaded the release, then maybe it is worth doing?)
